### PR TITLE
feat(operator): add topology label controller for node label copy

### PR DESCRIPTION
**Pull request by tmonty12**
## Summary
- Add `TopologyLabelReconciler` to watch annotated, scheduled pods and copy the configured topology label from their node.
- Mark worker DCDs with `nvidia.com/topology-label-key` from `spec.experimental.kvTransferPolicy.labelKey` and propagate it onto pod annotations.
- Register the controller and generate the minimal pod patch / node get RBAC needed for the copy path.
- Add dedicated Helm `topology-label-rbac.yaml` node-read ClusterRole/Binding so platform installs work in both cluster-wide and namespace-restricted modes.
- Tighten pod update predicates to enqueue only when topology label copy becomes newly actionable: scheduling, annotation add/change, or copied label removal.

## Validation
- `make manifests SHELL=/bin/sh .SHELLFLAGS=-ec`
- `KUBEBUILDER_ASSETS="$(bin/setup-envtest-release-0.19 use 1.29.0 --bin-dir "$(pwd)/bin" -p path)" go test ./internal/controller ./internal/dynamo ./cmd`
- `KUBEBUILDER_ASSETS="$(bin/setup-envtest-release-0.19 use 1.29.0 --bin-dir "$(pwd)/bin" -p path)" go test $(go list ./... | grep -v /e2e | grep -v /test | grep -v /cmd)`
- `go vet ./...`
- `helm template dyn ./deploy/helm/charts/platform/components/operator --set discoveryBackend=kubernetes --show-only templates/topology-label-rbac.yaml`
- `helm template dyn ./deploy/helm/charts/platform/components/operator --set discoveryBackend=kubernetes --set namespaceRestriction.enabled=true --show-only templates/topology-label-rbac.yaml`
- `helm template dyn ./deploy/helm/charts/platform/components/operator --set discoveryBackend=kubernetes --show-only templates/manager-rbac.yaml`
- `helm template dyn ./deploy/helm/charts/platform/components/operator --set discoveryBackend=kubernetes --set namespaceRestriction.enabled=true --show-only templates/manager-rbac.yaml`

## Links
- Linear: https://linear.app/nvidia/issue/DYN-3070/tar-pr4c-topology-label-controller-for-nodepod-label-copy
- DEP: https://github.com/ai-dynamo/dynamo/issues/9118

## Release Notes

* **New Features**
  * Added topology label controller that automatically copies node topology labels to worker pods when topology label configuration is enabled.

* **Tests**
  * Added comprehensive test suite validating topology label copying behavior and edge cases.

* **Chores**
  * Updated RBAC permissions to grant operator read access to cluster nodes and pod patching capabilities.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-dynamo/dynamo/pull/9879?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: deploy/operator/internal/controller/topology_label_controller.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	// Already has the label.
+	if _, exists := pod.Labels[labelKey]; exists {
+		return ctrl.Result{}, nil
+	}
+
+	// Not yet scheduled; reconcile again when spec.nodeName is set.
+	if pod.Spec.NodeName == "" {
```
Comment (julienmancuso) [inline deploy/operator/internal/controller/topology_label_controller.go]: nit: 
can we somehow make the “does this pod need a topology label copy?” helper return the label key:

```
labelKey, ok := topologyLabelCopyTarget(&pod)
if !ok {
	return ctrl.Result{}, nil
}
```

With a helper like:

```
func topologyLabelCopyTarget(pod *corev1.Pod) (string, bool) {
	labelKey := pod.GetAnnotations()[consts.KubeAnnotationTopologyLabelKey]
	if labelKey == "" {
		return "", false
	}
	if _, exists := pod.GetLabels()[labelKey]; exists {
		return "", false
	}
	if pod.Spec.NodeName == "" {
		return "", false
	}
	return labelKey, true
}
```

Then needsTopologyLabelCopy can call the same helper after the type assertion:

```
func needsTopologyLabelCopy(obj client.Object) bool {
	pod, ok := obj.(*corev1.Pod)
	if !ok || pod == nil {
		return false
	}

	_, ok = topologyLabelCopyTarget(pod)
	return ok
}
```

That would collapse the checks around lines 43-55 of deploy/operator/internal/controller/topology_label_controller.go into one named concept, while keeping the node lookup / missing-node-label / patch logic explicit.

File: deploy/operator/config/rbac/role.yaml [hunk 2]
```diff
@@ -21,6 +21,12 @@ rules:
   - patch
   - update
   - watch
+- apiGroups:
+  - ""
+  resources:
+  - nodes
+  verbs:
+  - get
```
Comment (julienmancuso) [inline deploy/operator/config/rbac/role.yaml:29]: you also need list/watch , a first read for Node starts/syncs a Node informer, which needs list/watch, not just get

File: deploy/operator/internal/controller/topology_label_controller.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	logger.Info("Copied node topology label to pod",
+		"pod", req.NamespacedName, "node", pod.Spec.NodeName,
+		"label", labelKey, "value", labelValue)
+	return ctrl.Result{}, nil
+}
+
+func (r *TopologyLabelReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithEventFilter(topologyLabelPredicate()).
```
Comment (julienmancuso) [inline deploy/operator/internal/controller/topology_label_controller.go]: we need this :
```
	WithEventFilter(predicate.And(
			commonController.EphemeralDeploymentEventFilter(r.Config, r.RuntimeConfig),
			topologyLabelPredicate(),
		)).
```

and 

```
if err = (&controller.TopologyLabelReconciler{
	Client:        mgr.GetClient(),
	Config:        operatorCfg,
	RuntimeConfig: runtimeConfig,
}).SetupWithManager(mgr); err != nil {
	return fmt.Errorf("unable to create TopologyLabel controller: %w", err)
}
```

reason: In namespace-restricted mode it keeps only the restricted namespace and in cluster-wide mode it respects excluded namespaces and restricted namespace filtering.

File: deploy/operator/internal/controller/topology_label_controller.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	var node corev1.Node
+	if err := r.Get(ctx, types.NamespacedName{Name: pod.Spec.NodeName}, &node); err != nil {
+		return ctrl.Result{}, fmt.Errorf("get node %s: %w", pod.Spec.NodeName, err)
+	}
+
+	labelValue, exists := node.Labels[labelKey]
+	if !exists {
+		logger.Info("Node missing topology label, skipping",
```
Comment (julienmancuso) [inline deploy/operator/internal/controller/topology_label_controller.go:63]: If these labels are guaranteed to exist before scheduling, then fine.
however if the pod is scheduled before the node label appears, the pod can remain without topology metadata forever. 
maybe add a bounded RequeueAfter or watch node label updates and enqueue pods on that node.

File: deploy/operator/cmd/main.go [hunk 5]
```diff
@@ -712,6 +712,12 @@ func registerControllers(
 		}
 	}
 
+	if err = (&controller.TopologyLabelReconciler{
+		Client: mgr.GetClient(),
+	}).SetupWithManager(mgr); err != nil {
+		return fmt.Errorf("unable to create TopologyLabel controller: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (tmonty12) [reply deploy/operator/cmd/main.go:723]: I went with the approach of using the `APIReader` client for getting the individual node a pod is scheduled on instead of continuing to use the cached client which which would require `list/watch` perms for the informer. This follows the DGDR reconciler pattern for gpu discovery and avoids the memory overhead of storing all node objects.

File: deploy/operator/config/rbac/role.yaml (hunk 2 above)
Comment (tmonty12) [reply deploy/operator/config/rbac/role.yaml:29]: No longer, needed. See comment here: https://github.com/ai-dynamo/dynamo/pull/9879#discussion_r3289925220

File: deploy/operator/internal/dynamo/graph.go [hunk 6]
```diff
@@ -379,6 +379,14 @@ func generateSingleDCD(
 
 	applyDGDTemplateDefaults(&deployment.Spec.DynamoComponentDeploymentSharedSpec, parentDGD)
 
+	// Topology label controller marker: set on the DCD so it propagates to pods.
+	if shouldApplyKvTransferPolicyToWorkerComponent(component, parentDGD) {
+		if deployment.Annotations == nil {
+			deployment.Annotations = make(map[string]string)
+		}
+		deployment.Annotations[commonconsts.KubeAnnotationTopologyLabelKey] = parentDGD.Spec.Experimental.KvTransferPolicy.LabelKey
+	}
```
[the remark this answers is not in the document]
Comment (tmonty12) [reply deploy/operator/internal/dynamo/graph.go:389]: good catch! the CRD invariant requires `labelKey` be non-empty today but expansion to Grove ClusterTopology could change this invariant in the future. adding guard to only set the topology label annotation is `labelKey != ""`

File: deploy/operator/internal/controller/topology_label_controller.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+)
+
+// TopologyLabelReconciler watches worker pods that have the
+// nvidia.com/topology-label-key annotation. When a pod is scheduled
+// (spec.nodeName is set) but the target label is missing, the controller
+// reads the label from the node and patches it onto the pod. The Downward
+// API volume then picks up the label value.
+type TopologyLabelReconciler struct {
+	client.Client
+}
```
[the remark this answers is not in the document]
Comment (tmonty12) [reply deploy/operator/internal/controller/topology_label_controller.go:39]: in `graph.go` we filter on only worker components to add `nvidia.com/topology-label-key`. I've added an additional boundary `isDynamoComponentPod` for extra measure but I don't want to duplicate the filtering in `graph.go` and the topology label controller.

[the remark this answers is not in the document]
Comment (tmonty12) [reply deploy/operator/internal/controller/topology_label_controller.go:39]: check now

File: deploy/operator/internal/controller/topology_label_controller.go (hunk 1 above)
Comment (tmonty12) [reply deploy/operator/internal/controller/topology_label_controller.go]: done!

File: deploy/operator/internal/controller/topology_label_controller.go (hunk 3 above)
Comment (tmonty12) [reply deploy/operator/internal/controller/topology_label_controller.go]: good catch! done

File: deploy/operator/internal/dynamo/graph.go [hunk 8]
```diff
@@ -379,6 +379,14 @@ func generateSingleDCD(
 
 	applyDGDTemplateDefaults(&deployment.Spec.DynamoComponentDeploymentSharedSpec, parentDGD)
 
+	// Topology label controller marker: set on the DCD so it propagates to pods.
+	if shouldApplyKvTransferPolicyToWorkerComponent(component, parentDGD) {
```
Comment (dynamo-ops) [inline deploy/operator/internal/dynamo/graph.go:384]: The marker is written only to generated DCD metadata, so Grove-rendered worker PodClique pods get the KV topology env/DownwardAPI volume without the controller annotation and will never have the node label copied. Fix: add the topology-label annotation to non-GMS Grove worker clique annotations when the same KvTransferPolicy applies.

<details><summary>🤖 AI Fix</summary>

In deploy/operator/internal/dynamo/graph.go, update buildCliqueForRole after generateAnnotations(p.component, p.dynamoDeployment, p.componentName) succeeds and before clique.Annotations = annotations to set annotations[commonconsts.KubeAnnotationTopologyLabelKey] = p.dynamoDeployment.Spec.Experimental.KvTransferPolicy.LabelKey when p.r.Role != RoleGMS && shouldApplyKvTransferPolicyToWorkerComponent(p.component, p.dynamoDeployment), and add a GenerateGrovePodCliqueSet unit test proving worker/decode/prefill cliques with KvTransferPolicy carry the annotation while RoleGMS cliques do not.

</details>

File: deploy/operator/internal/controller/topology_label_controller.go (hunk 4 above)
Comment (tmonty12) [reply deploy/operator/internal/controller/topology_label_controller.go:63]: This is a good point. After some thinking, I think it's fine to keep the logic as is - no requeue. This is because if the node label dne, then the pod never receives the copied label, the runtime will never receive the domain/value via the downward API and will exit after 30s. The Pod will then be rescheduled. This should flag to the user that the node label dne. 

You can check out the python/runtime logic here: https://github.com/ai-dynamo/dynamo/pull/9878

Watching on all node label updates would add considerable overhead and doing a bounded RequeueAfter would require more logic for using annotations to track when the operator first saw the pod scheduled on a node with no label.

I'll add a Pod event `TopologyLabelMissing` that makes it clear to kubernetes operators that the node does not have the required topology label.

File: deploy/operator/internal/dynamo/graph.go (hunk 8 above)
Comment (tmonty12) [reply deploy/operator/internal/dynamo/graph.go:384]: great catch! fixed

File: deploy/operator/internal/controller/topology_label_controller.go (hunk 4 above)
Comment (julienmancuso) [reply deploy/operator/internal/controller/topology_label_controller.go:63]: yup, makes sense

**Review by julienmancuso — APPROVED** [review]
